### PR TITLE
Remix Vite plugin support

### PR DESCRIPTION
**Pull request by TooTallNate**
Adds support for Remix apps which use the new Remix Vite plugin.

* The vanilla Remix + Vite template deploys correctly out-of-the-box, however only a single Node.js function will be used, and a warning will be printed saying to configure the `vercelPreset()` Preset.
* When used in conjunction with the `vercelPreset()` Preset (https://github.com/vercel/remix/pull/81), allows for the application to utilize Vercel-specific features, like per-route `export const config` configuration, including multi-runtime (Node.js / Edge runtimes) within the same app.

## To test this today

1. Generate a Remix + Vite project from the template:
    ```
    npx create-remix@latest --template remix-run/remix/templates/vite
    ```
1. Install `@vercel/remix`:
    ```
    npm i --save-dev @vercel/remix
    ```
1. Ensure you are using Remix v2.8.0 or later.
1. Configure the `vercelPreset()` in the `vite.config.ts` file:
    ```diff
    --- a/vite.config.ts
    +++ b/vite.config.ts
    @@ -1,10 +1,11 @@
     import { vitePlugin as remix } from "@remix-run/dev";
     import { installGlobals } from "@remix-run/node";
     import { defineConfig } from "vite";
    +import { vercelPreset } from "@vercel/remix/vite";
     import tsconfigPaths from "vite-tsconfig-paths";
    
     installGlobals();
    
     export default defineConfig({
    -  plugins: [remix(), tsconfigPaths()],
    +  plugins: [remix({ presets: [vercelPreset()] }), tsconfigPaths()],
     });
    ```
1. Create a new Vercel Project in the dashboard, and ensure the Vercel preset is set to "Remix" in the Project Settings. The autodetection will work correctly once this PR is merged, but for now it gets incorrectly detected as "Vite" preset.
    * **Hint**: You can create a new empty Project by running the `vercel link` command.
    Screenshot 2024-02-27 at 10 37 11
1. Deploy to Vercel:
    ```
    vercel deploy
    ```

### CODE REVIEW

File: packages/remix/test/fixtures-vite/01-no-preset/vercel.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+  "framework": "remix"
```
Comment (TooTallNate) [inline packages/remix/test/fixtures-vite/01-no-preset/vercel.json:2]: Remix + Vite is currently detected as `vite` preset. The changes in this PR fix that, but these changes need to be deployed on the API-side first. So we can remove these `vercel.json` files in the fixtures in a follow-up PR after this PR is deployed.

File: packages/remix/test/fixtures-vite/02-interactive-remix-routing-v2/package.json [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    "@remix-run/node": "2.7.2",
+    "@remix-run/react": "2.7.2",
+    "@remix-run/serve": "2.7.2",
+    "@vercel/remix": "2.7.2-preset.1",
+    "isbot": "4.4.0",
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0"
+  },
+  "devDependencies": {
+    "@remix-run/dev": "https://remix-e7asp269z.vercel.sh/remix-dev.tgz",
```
Comment (TooTallNate) [inline packages/remix/test/fixtures-vite/02-interactive-remix-routing-v2/package.json:26]: **Note:** We'll want to update this to 2.7.3 once that version is released.

**Comment by depsimon** [comment]
> * The vanilla Remix + Vite template deploys correctly out-of-the-box

Hi @TooTallNate can you point out the vanilla Remix + Vite template you used?

I tried with `create-remix@latest --template remix-run/remix/templates/vite` but this does not deploy correctly out-of-the-box on Vercel (deploys correctly but getting 404 after configuring the output directory to `/build/client`).

Did you use a specific config to make it work?

**Comment by k1sul1** [comment]
Great work on this! Any chance of getting an ETA on when this could be available in production? 

Next week? Next month? Any info would make it much easier to decide on how we should proceed with a blocked PR and will be greatly appreciated!

**Comment by TooTallNate** [comment]
@depsimon I've added some instructions at the top. Let me know if you bump into any issues.

@k1sul1 We're aiming for this week. Remix v2.8.0 also needs to be released as it contains a bug fix related to the server bundles feature (but I believe that is happening in the next couple of days).

**Comment by ToppleTheNun** [comment]
After following the instructions in the PR body, I'm getting these errors on my test repo:
```
Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/var/task/build/index.js' imported from /var/task/build/server-index.mjs
    at new NodeError (node:internal/errors:405:5)
    at finalizeResolution (node:internal/modules/esm/resolve:332:11)
    at moduleResolve (node:internal/modules/esm/resolve:1029:10)
    at moduleResolveWithNodePath (node:internal/modules/esm/resolve:876:12)
    at defaultResolve (node:internal/modules/esm/resolve:1242:79)
    at ModuleLoader.defaultResolve (node:internal/modules/esm/loader:403:12)
    at ModuleLoader.resolve (node:internal/modules/esm/loader:372:25)
    at ModuleLoader.getModuleJob (node:internal/modules/esm/loader:249:38)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:76:39)
    at link (node:internal/modules/esm/module_job:75:36) {
  url: 'file:///var/task/build/index.js',
  code: 'ERR_MODULE_NOT_FOUND'
}
INIT_REPORT Init Duration: 325.29 ms	Phase: invoke	Status: error	Error Type: Runtime.ExitError
Error: Runtime exited with error: exit status 1
```

Minimal reproduction: https://github.com/ToppleTheNun/remix-vite-vercel

**Comment by TooTallNate** [comment]
@ToppleTheNun Could you share the deployment URL? It deployed correctly for me: https://remix-vite-vercel-epvyuy9p4-uncurated-tests.vercel.app/

**Comment by ToppleTheNun** [comment]
@TooTallNate

Deployment URL: https://vercel.com/topplethenun/remix-vite-vercel/G8UPjGaGrdPJ4FA8uytXBWy8RDp2
Deployed URL: https://remix-vite-vercel-ff0uj5th5-topplethenun.vercel.app/

**Comment by TooTallNate** [comment]
Looks like you missed the `VERCEL_CLI_VERSION` env var part, so the deployment was built using the wrong version of the CLI (which doesn't contain the changes in this PR).

**Comment by ToppleTheNun** [comment]
I ran the deploy from my local with this:
```shell
$ vercel deploy -b VERCEL_CLI_VERSION=https://vercel-git-tootallnate-zero-1217-research-remix-v-next-vite.vercel.sh/tarballs/vercel.tgz
```

**Comment by ToppleTheNun** [comment]
Ah, I see where it went wrong. I clicked the "promote to production" link in the Vercel console and it built without using the CLI. That'd do it. Apologies!

**Comment by TooTallNate** [comment]
Screenshot 2024-02-27 at 14 44 15

You should see this log line in the deployment build logs (which I don't see for your deployment). So looks like it didn't take for some reason. Perhaps try setting the env var in the Project Settings on the Vercel dashboard instead.

**Comment by justyn-clark** [comment]
Screenshot 2024-02-27 at 2 34 27 PM


Screenshot 2024-02-27 at 2 39 38 PM

------------

This deployment was successful pushing the branch to Github
https://remix-vite-vercel-test-dvfhsdxv6-justyn-clark.vercel.app

This deployment was successful using vercel deploy from step 6. above
https://remix-vite-vercel-test-bg8ziwstj-justyn-clark.vercel.app

```
vercel deploy -b VERCEL_CLI_VERSION=https://vercel-git-tootallnate-zero-1217-research-remix-v-next-vite.vercel.sh/tarballs/vercel.tgz
```

------------

Out of the box I'm a happy camper. Nice job and thanks for giving this attention.  🎉

**Comment by depsimon** [comment]
> @depsimon I've added some instructions at the top. Let me know if you bump into any issues.
> 
> @k1sul1 We're aiming for this week. Remix v2.8.0 also needs to be released as it contains a bug fix related to the server bundles feature (but I believe that is happening in the next couple of days).

Thanks @TooTallNate, works fine following your instructions 🦾

**Comment by krall12** [comment]
Following the steps with the custom VERCEL_CLI_VERSION I'm getting a bun install error. 

Screenshot 2024-02-28 at 10 57 57 AM

File: packages/remix/src/build-vite.ts [hunk 3]
```diff
@@ -0,0 +1,477 @@
...
+    } else {
+      await execCommand('remix build', {
+        ...spawnOpts,
+        cwd: entrypointFsDirname,
+      });
+    }
+  }
+
+  // This needs to happen before we run NFT to create the Node/Edge functions
+  // TODO: maybe remove this?
```
Comment (EndangeredMassa) [inline packages/remix/src/build-vite.ts:161]: **question (soft-blocking):** should we remove this?‏

**Review by EndangeredMassa — APPROVED** [review]
**suggestion (soft-blocking):** add unit tests

The logic overall seems good, but there's a lot of it, making it harder to review as a whole. I think we should break more of that into testable chunks at the unit level.

**Pull request merged**